napari/layers/volume/volume.py

```diff
@@ -0,0 +1,245 @@
+
+        self.events.add(clim=Event, colormap=Event, rendering=Event)
+
+        with self.freeze_refresh():
+            # Set data
+            self._data = volume
+            self.metadata = metadata or {}
+
+            # Intitialize volume views and thumbnails with zeros
+            self._data_view = np.zeros((1, 1, 1))
+            self._data_thumbnail = self._data_view
+
+            # Set clims and colormaps
+            self._colormap_name = ''
+            self._clim_msg = ''
+            if clim_range is None:
+                self._clim_range = calc_data_range(self.data)
+            else:
+                self._clim_range = clim_range
+
+            if clim is None:
+                self.clim = self._clim_range
+            else:
+                self.clim = clim
+            self.colormap = colormap
+
+            # Set update flags
+            self._need_display_update = False
+            self._need_visual_update = False
+
+            # Re intitialize indices
+            self._indices = (0,) * (self.ndim - 3) + (
+                slice(None, None, None),
+                slice(None, None, None),
+                slice(None, None, None),
+            )
+
+            # Trigger generation of view slice and thumbnail
+            self._set_view_slice()
+
+    @property
+    def data(self):
+        """array: Image data."""
+        return self._data
+
+    @data.setter
+    def data(self, data):
+        self._data = data
+        self.events.data()
+        self.refresh()
+
+    def _get_shape(self):
+        return self.data.shape
+
+    @property
+    def colormap(self):
+        """2-tuple of str, vispy.color.Colormap: colormap for luminance images.
+        """
+        return self._colormap_name, self._node.cmap
+
+    @colormap.setter
+    def colormap(self, colormap):
+        name = '[unnamed colormap]'
+        if isinstance(colormap, str):
+            name = colormap
+        elif isinstance(colormap, tuple):
+            name, cmap = colormap
+            self._colormaps[name] = cmap
+        elif isinstance(colormap, dict):
+            self._colormaps.update(colormap)
+            name = list(colormap)[0]  # first key in dict
+        elif isinstance(colormap, vispy.color.Colormap):
+            name = increment_unnamed_colormap(
+                name, list(self._colormaps.keys())
+            )
+            self._colormaps[name] = colormap
+        else:
+            warnings.warn(f'invalid value for colormap: {colormap}')
+            name = self._colormap_name
+        self._colormap_name = name
+        self._node.cmap = self._colormaps[name]
+        self.events.colormap()
+
+    @property
+    def colormaps(self):
+        """tuple of str: names of available colormaps."""
+        return tuple(self._colormaps.keys())
+
+    def _set_view_slice(self):
+        """Set the view given the indices to slice with."""
+        self._data_view = np.asarray(self.data[tuple(self._indices)])
```

Comment:
This needs to be
```python
 indices = list(self.indices)
        indices[:-3] = np.clip(
            indices[:-3], 0, np.subtract(self.shape[:-3], 1)
        )
self._data_view = np.asarray(self.data[tuple(self.indices)])
```
otherwise you may get errors for data > 3D

Reply:
if you try and slice into data that is larger than the shape